overlord/snapstate/snapmgr.go

```diff
@@ -640,8 +640,29 @@ func (m *SnapManager) doDiscardSnap(t *state.Task, _ *tomb.Tomb) error {
 	return nil
 }
 
+type cachedUpdateRevisionsWasRun struct{}
+
 // Ensure implements StateManager.Ensure.
 func (m *SnapManager) Ensure() error {
+	// ensure the core/kernel revisions that we booted with match
+	// the active revision on disk (they may get out of sync on
+	// a failed boot that falls back to the previous version for
+	// example)
+	m.state.Lock()
+	wasRun := m.state.Cached(cachedUpdateRevisionsWasRun{})
+	m.state.Unlock()
+
+	if wasRun == nil {
+		m.state.Lock()
+		err := UpdateRevisions(m.state)
+		m.state.EnsureBefore(0)
```

Comment:
Can we let UpdateRevisions itself do that if actually necessary?  Otherwise we're assuming we know what happened there.